### PR TITLE
Add support for pm-cpu

### PR DESCRIPTION
**Pull request by xylar**
This is the machine name for Perlmutter's CPU nodes at NERSC.

### CODE REVIEW

**Comment by xylar** [comment]
I'm going to rebase this in a few minutes once #61 gets merged.

**Comment by xylar** [comment]
@forsyth2, I'm going to test this in my packages compass and MPAS-Analysis to make sure it works as expected tomorrow.  But a quick build, install and test worked as expected:
```
conda mambabuild conda
mamba create -y -n mache_dev --use-local mache=1.7.0
mamba activate mache_dev
python -c "import mache; print(mache.MachineInfo())"
```

**Comment by forsyth2** [comment]
@xylar I can't get `mache` installed locally. I'm having difficulty getting `mamba` set up and I can't get it working just using `conda`. In any case, Perlmutter is now down for maintenance until tomorrow.
```
              Perlmutter is going down for maintenance from                    
               2022-10-24T15:00 PDT - 2022-10-25T12:00 PDT  
```

**Comment by xylar** [comment]
Can you be more specific about what didn't work? Did you install both mamba and boa (the latter provides mambabuild)?

**Comment by forsyth2** [comment]
Trying on Cori:
```
$ conda install -y mamba # worked on Cori, not on Perlmutter
$ conda mambabuild conda

CommandNotFoundError: No command 'conda mambabuild'.
Did you mean 'conda build'?
$ conda install -y boa # worked on Cori, didn't try on Perlmutter
$ mamba create -y -n mache_dev --use-local mache=1.7.0 # Failed on both Cori and Perlmutter
Looking for: ['mache=1.7.0']

pkgs/r/linux-64                                               No change
pkgs/main/noarch                                              No change
pkgs/r/noarch                                                 No change
pkgs/main/linux-64                                   4.9MB @   4.7MB/s  1.1s
conda-forge/noarch                                  10.0MB @   4.9MB/s  2.2s
conda-forge/linux-64                                26.4MB @   5.4MB/s  5.3s
Encountered problems while solving:
  - nothing provides requested mache 1.7.0**
```

**Comment by forsyth2** [comment]
Trying on Cori after installing `boa` works. I'll try again when Perlmutter is back up.
```
$ conda install -y mamba # worked on Cori, not on Perlmutter
$ conda mambabuild conda

CommandNotFoundError: No command 'conda mambabuild'.
Did you mean 'conda build'?
$ conda install -y boa # worked on Cori, didn't try on Perlmutter
$ conda mambabuild conda # Now works, with boa
$ mamba create -y -n mache_dev --use-local mache=1.7.0 # works, failed on Perlmutter
$ mamba activate mache_dev
Run 'mamba init' to be able to run mamba activate/deactivate
and start a new shell session. Or use conda to activate/deactivate.
$ conda activate mache_dev
$ python -c "import mache; print(mache.MachineInfo())" # Works
```

**Comment by forsyth2** [comment]
@xylar I just ran the following on Perlmutter:
```
$ conda activate mache_dev # Still available from creating on Cori a few days ago
$ python -c "import mache; print(mache.MachineInfo())"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/global/u1/f/forsyth/mache/mache/machine_info.py", line 75, in __init__
    raise ValueError('Unable to discover machine form host name')
ValueError: Unable to discover machine form host name
```
Although, that's not running on your latest commit for this pull request. (Also, note the error should probably be "from host name", not "form host name").

I also need to complete https://github.com/E3SM-Project/zppy/issues/310 before I can confirm for sure that `zppy` works fine on Perlmutter using `mache`.

**Comment by xylar** [comment]
@forsyth2, I think this indicates that you are not running with this version of mache, as you suggested.  I'm happy to fix that typo in this PR.  Please try either rebuilding the local package with this branch or installing with `python -m pip install -e .` to get the version of mache from this branch.  Please report back if you still have the problem when you do.

**Comment by xylar** [comment]
@forsyth2, can you also do:
```
env | grep perlmutter
```
for me?  I am using an environment variable (`$LMOD_SYSTEM_NAME`) because that was the only way I could figure out to determine that I'm on `perlmutter` but if you don't have that variable, it's not a reliable way.

**Comment by xylar** [comment]
For example:
```
> echo $NERSC_HOST
unknown
```

**Comment by forsyth2** [comment]
@xylar Thanks, that got it working.
```
$ git remote add xylar git@github.com:xylar/mache.git
$ git fetch xylar
$ git checkout -b add_pm-cpu xylar/add_pm-cpu
$ python -m pip install -e .
$ python -c "import mache; print(mache.MachineInfo())"
/global/u1/f/forsyth/mache/mache/discover.py:42: UserWarning: defaulting to pm-cpu.  Explicitly specify pm-gpu as the machine if you wish to run on GPUs.
  warnings.warn('defaulting to pm-cpu.  Explicitly specify pm-gpu '
Machine: pm-cpu
  E3SM Supported Machine: False

E3SM-Unified:
  E3SM-Unified is not currently loaded
  Base path: /global/common/software/e3sm/anaconda_envs

Diagnostics:
  Base path: /global/cfs/cdirs/e3sm/diagnostics

Config options:
  [e3sm_unified]
    group = e3sm
    compiler = gnu
    mpi = mpich
    base_path = /global/common/software/e3sm/anaconda_envs

  [diagnostics]
    base_path = /global/cfs/cdirs/e3sm/diagnostics
    group = e3sm

  [web_portal]
    base_path = /global/cfs/cdirs/e3sm/www
    base_url = https://portal.nersc.gov/cfs/e3sm
    username = forsyth

  [parallel]
    system = slurm
    parallel_executable = srun
    cores_per_node = 256
    account = e3sm
    constraints = cpu
    qos = regular, premium, debug

  [spack]
    modules_before = False
    modules_after = False

  [sync]
    hostname = perlmutter-p1.nersc.gov
```

**Comment by xylar** [comment]
@forsyth2, can you share how `$NERSC_HOST` is getting set for you?  It doesn't seem like a reliable thing because it's not working for me but at least I'd like to know.

**Comment by forsyth2** [comment]
I didn't do anything to set `$NERSC_HOST` and I'm not seeing any obvious place where it gets set.

**Comment by forsyth2** [comment]
@xylar Why is the constraint included as part of the machine name? I know you mentioned this with Cori too on https://github.com/E3SM-Project/zppy/pull/330#discussion_r1002382547.
```
$ python -c "import mache; print(mache.MachineInfo().machine)"
pm-cpu
$ python -c "import mache; print(mache.MachineInfo().get_account_defaults())"
('e3sm', None, 'cpu', 'regular')
```
It seems like unnecessary duplication of information. In `zppy`, we often just checked for `cori` -- if we use the `mache` convention we would have to check for two different machines (`cori-haswell, cori-knl` or `pm-cpu, pm-gpu`)

**Comment by xylar** [comment]
@forsyth2, okay, regardin `$NERSC_HOST`, maybe my `.bash_profile.ext` is messing thing up and the default value would be useful.

**Comment by xylar** [comment]
> Why is the constraint included as part of the machine name?

This is what is done in E3SM and `mache` is trying to be exactly consistent with E3SM:
https://github.com/E3SM-Project/E3SM/blob/master/cime_config/machines/config_machines.xml#L162

For E3SM, there are so many differences between `pm-gpu` and `pm-cpu` (just as there were between `cori-haswell` and `cori-knl`) that they are basically separate machines.

**Comment by xylar** [comment]
>  if we use the `mache` convention we would have to check for two different machines (`cori-haswell, cori-knl` or `pm-cpu, pm-gpu`)

Yes, that's exactly what I would advocate.  It is worth the potential redundancy.

**Comment by forsyth2** [comment]
> For E3SM, there are so many differences between `pm-gpu` and `pm-cpu` (just as there were between `cori-haswell` and `cori-knl`) that they are basically separate machines.

Interesting. Thanks for explaining. That will help with implementing https://github.com/E3SM-Project/zppy/issues/310.

**Comment by xylar** [comment]
@forsyth2, okay, I got the host detection thing cleared up at NERSC.  I should, indeed use `$NERSC_HOST`.  I see that now that I deleted my old `.bash_profile.ext` file (and similar for other shells).  I added an error message when `$NERSC_HOST=unknown` is found that will hopefully help users in the same situation I was in.

**Comment by xylar** [comment]
@forsyth2, this is working for me in compass, which does some pretty heavy testing (including of the spack files as well as the config options and machine discovery).  I'm pretty confident it's working.

Could you do one or two last tests and either approve or request changes?  I'd like to merge this an make the 1.7.0 release by the end of the day if possible, or early tomorrow morning if necessary.

File: mache/discover.py [hunk 1]
```diff
@@ -27,11 +29,26 @@ def discover_machine():
...
+            warnings.warn('defaulting to pm-cpu.  Explicitly specify '
+                          'pm-gpu as the machine if you wish to run on '
+                          'GPUs.')
+            machine = 'pm-cpu'
+        elif hostname == 'unknown':
+            raise ValueError(
+                'You appear to have $NERSC_HOST=unknown.  This typically '
+                'indicates that you \n'
+                'have an outdated .bash_profile.ext or similar.  Please '
+                'delete that file, \n'
```
Comment (forsyth2) [inline mache/discover.py:52]: I have a lot of useful stuff in my `~/.bash_profile.ext`, so I wouldn't recommend just deleting that file no matter what. Perhaps replace "Please delete that file" with "Please edit that file or delete if necessary".

**Review by forsyth2 — APPROVED** [review]
I think this is good to merge. Thanks @xylar!

Tested with:
```
$ cd mache
$ git fetch xylar
$ git checkout -b add_pm-cpu2 xylar/add_pm-cpu
$ python -m pip install -e .
$ python -c "import mache; print(mache.MachineInfo())"
/global/u1/f/forsyth/mache/mache/discover.py:43: UserWarning: defaulting to pm-cpu.  Explicitly specify pm-gpu as the machine if you wish to run on GPUs.
  warnings.warn('defaulting to pm-cpu.  Explicitly specify '
Machine: pm-cpu
  E3SM Supported Machine: False

E3SM-Unified:
  E3SM-Unified is not currently loaded
  Base path: /global/common/software/e3sm/anaconda_envs

Diagnostics:
  Base path: /global/cfs/cdirs/e3sm/diagnostics

Config options:
  [e3sm_unified]
    group = e3sm
    compiler = gnu
    mpi = mpich
    base_path = /global/common/software/e3sm/anaconda_envs

  [diagnostics]
    base_path = /global/cfs/cdirs/e3sm/diagnostics
    group = e3sm

  [web_portal]
    base_path = /global/cfs/cdirs/e3sm/www
    base_url = https://portal.nersc.gov/cfs/e3sm
    username = forsyth

  [parallel]
    system = slurm
    parallel_executable = srun
    cores_per_node = 256
    account = e3sm
    constraints = cpu
    qos = regular, premium, debug

  [spack]
    modules_before = False
    modules_after = False

  [sync]
    hostname = perlmutter-p1.nersc.gov

$ python -c "import mache; print(mache.MachineInfo().machine)"
/global/u1/f/forsyth/mache/mache/discover.py:43: UserWarning: defaulting to pm-cpu.  Explicitly specify pm-gpu as the machine if you wish to run on GPUs.
  warnings.warn('defaulting to pm-cpu.  Explicitly specify '
pm-cpu
$ python -c "import mache; print(mache.MachineInfo().get_account_defaults())"
/global/u1/f/forsyth/mache/mache/discover.py:43: UserWarning: defaulting to pm-cpu.  Explicitly specify pm-gpu as the machine if you wish to run on GPUs.
  warnings.warn('defaulting to pm-cpu.  Explicitly specify '
('e3sm', None, 'cpu', 'regular')
```

See https://github.com/E3SM-Project/zppy/pull/333 for further comments on testing.

**Comment by xylar** [comment]
Okay, I can rephrase. I think NERSC recommends against using the .ext files here https://docs.nersc.gov/environment/

**Comment by xylar** [comment]
Thanks for your testing @forsyth2!

**Comment by forsyth2** [comment]
> I think NERSC recommends against using the .ext files here https://docs.nersc.gov/environment/

Interesting, I guess I should move over my aliases and functions to `~/.bash_profile` directly then.

> Thanks for your testing @forsyth2!

No problem, thanks for setting up support for Perlmutter. This will be very useful.

**Pull request merged**